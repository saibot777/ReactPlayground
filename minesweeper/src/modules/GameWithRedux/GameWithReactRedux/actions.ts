import { AnyAction } from "redux";
import { Coords } from "@/core/Field";

// Actions
export const OPEN_CELL = "modules/GameWithRedux/OPEN_CELL";

// Action Creators
export const openCell = (coords: Coords): AnyAction => ({
  type: OPEN_CELL,
  payload: { coords },
});
