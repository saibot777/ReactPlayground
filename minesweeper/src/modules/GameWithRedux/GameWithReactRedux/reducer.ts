import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import {
  Field,
  CellState,
  generateFieldWithDefaultState,
  fieldGenerator,
  Coords,
} from "@/core/Field";
import { LevelNames, GameSettings } from "@/modules/GameSettings";
import { openCell as openCellHandler } from "@/core/openCell";
import { setFlag } from "@/core/setFlag";

export interface State {
  level: LevelNames;
  time: number;
  bombs: number;
  isGameOver: boolean;
  isGameStarted: boolean;
  isWin: boolean;
  settings: [number, number];
  playerField: Field;
  gameField: Field;
  flagCounter: number;
}

export const getInitialState = (level = "beginner" as LevelNames): State => {
  const settings = GameSettings[level];
  const [size, bombs] = settings;

  return {
    level,
    time: 0,
    bombs,
    isGameOver: false,
    isGameStarted: false,
    isWin: false,
    settings,
    flagCounter: 0,
    playerField: generateFieldWithDefaultState(size, CellState.hidden),
    gameField: fieldGenerator(size, bombs / (size * size)),
  };
};

export const { reducer, actions } = createSlice({
  name: "game",
  initialState: getInitialState(),
  reducers: {
    openCell: (state, { payload }: PayloadAction<Coords>) => {
      const { playerField, gameField } = state;
      try {
        const [newPlayerField, isSolved] = openCellHandler(
          payload,
          playerField,
          gameField
        );
        state.isGameStarted = !isSolved;
        state.isGameOver = isSolved;
        state.isWin = isSolved;
        state.playerField = newPlayerField;
      } catch (error) {
        state.isGameStarted = false;
        state.isGameOver = true;
        state.isWin = false;
        state.playerField = gameField;
      }
    },
    setFlag: (state, { payload }: PayloadAction<Coords>) => {
      const { playerField, gameField, flagCounter, bombs } = state;

      const [newPlayerField, isSolved, newFlagCounter] = setFlag(
        payload,
        playerField,
        gameField,
        flagCounter,
        bombs
      );
      state.isGameStarted = !isSolved;
      state.isGameOver = isSolved;
      state.isWin = isSolved;
      state.flagCounter = newFlagCounter;
      state.playerField = newPlayerField;
    },
    reset: ({ level }) => getInitialState(level),
    changeLevel: (state, { payload }: PayloadAction<LevelNames>) =>
      getInitialState(payload),
  },
});
