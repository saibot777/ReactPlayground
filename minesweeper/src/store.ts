import { configureStore } from "@reduxjs/toolkit";

import { gameSlice } from "@/modules/GameWithRedux";

export const store = configureStore({
  reducer: {
    game: gameSlice.reducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
