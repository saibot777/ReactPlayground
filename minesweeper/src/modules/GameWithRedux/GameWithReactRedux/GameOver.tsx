import React, { FC, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { RootState } from '@/store';
import { GameOver as GameOverComponent } from '@/components/Game';
import { actions } from '@/modules/GameWithRedux';

export const GameOver: FC = () => {
  const { isGameOver, isWin } = useSelector(
    ({ game: { isGameOver, isWin } }: RootState) => ({
      isGameOver,
      isWin,
    })
  );

  const dispatch = useDispatch();

  const onReset = useCallback(
    () => dispatch(actions.reset()),
    // Stryker disable next-line ArrayDeclaration
    []
  );

  return (
    <>{isGameOver && <GameOverComponent onClick={onReset} isWin={isWin} />}</>
  );
};
