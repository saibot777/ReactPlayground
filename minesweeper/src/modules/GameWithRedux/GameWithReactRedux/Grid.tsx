import React, { FC, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { Coords } from '@/core/Field';
import { RootState } from '@/store';
import { Grid as GridComponent } from '@/components/Grid';

import { actions, runTimer } from '@/modules/GameWithRedux/game';

export const Grid: FC = () => {
  const { playerField } = useSelector(
    ({ game: { playerField } }: RootState) => ({
      playerField,
    })
  );

  const dispatch = useDispatch();

  const onClick = useCallback(
    (coords: Coords) => {
      dispatch(actions.openCell(coords));
      dispatch(runTimer());
    },
    // Stryker disable next-line ArrayDeclaration
    []
  );

  const onContextMenu = useCallback(
    (coords: Coords) => {
      dispatch(actions.setFlag(coords));
      dispatch(runTimer());
    },
    // Stryker disable next-line ArrayDeclaration
    []
  );

  return (
    <GridComponent onClick={onClick} onContextMenu={onContextMenu}>
      {playerField}
    </GridComponent>
  );
};
