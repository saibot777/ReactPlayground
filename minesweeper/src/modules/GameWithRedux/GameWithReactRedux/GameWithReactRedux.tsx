import React, { FC } from 'react';

import { Scoreboard } from './Scoreboard';
import { Grid } from './Grid';
import { GameOver } from './GameOver';

export const GameWithReactRedux: FC = () => {
  return (
    <>
      <Scoreboard />
      <GameOver />
      <Grid />
    </>
  );
};
