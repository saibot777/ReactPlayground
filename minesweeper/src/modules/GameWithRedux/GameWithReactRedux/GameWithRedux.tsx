import React, { FC, useReducer, useCallback } from 'react';

import { Coords } from '@/core/Field';

import { GameLevels, LevelNames } from '@/modules/GameSettings';

import { Scoreboard } from '@/components/Scoreboard';
import { Grid } from '@/components/Grid';
import { GameOver } from '@/components/Game';

import { reducer, actions, getInitialState } from './reducer';

export const GameWithRedux: FC = () => {
  const [
    { level, time, isGameOver, isWin, settings, playerField, flagCounter },
    dispatch,
  ] = useReducer(reducer, getInitialState());

  const [, bombs] = settings;

  const onClick = useCallback(
    (coords: Coords) => dispatch(actions.openCell(coords)),
    // Stryker disable next-line ArrayDeclaration
    []
  );

  const onReset = useCallback(
    () => dispatch(actions.reset()),
    // Stryker disable next-line ArrayDeclaration
    []
  );

  const onContextMenu = useCallback(
    (coords: Coords) => dispatch(actions.setFlag(coords)),
    // Stryker disable next-line ArrayDeclaration
    []
  );

  const onChangeLevel = useCallback(
    ({ target: { value: level } }: React.ChangeEvent<HTMLSelectElement>) =>
      dispatch(actions.changeLevel(level as LevelNames)),
    // Stryker disable next-line ArrayDeclaration
    []
  );

  return (
    <>
      <Scoreboard
        time={String(time)}
        bombs={String(bombs - flagCounter)}
        levels={GameLevels as unknown as string[]}
        defaultLevel={level}
        onChangeLevel={onChangeLevel}
        onReset={onReset}
      />
      {isGameOver && <GameOver onClick={onReset} isWin={isWin} />}
      <Grid onClick={onClick} onContextMenu={onContextMenu}>
        {playerField}
      </Grid>
    </>
  );
};