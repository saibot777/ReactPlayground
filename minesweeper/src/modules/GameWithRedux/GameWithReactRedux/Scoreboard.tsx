import React, { FC, useEffect, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useSearchParams } from 'react-router-dom';

import { RootState } from '@/store';
import { GameLevels, LevelNames } from '@/modules/GameSettings';
import { Scoreboard as ScoreboardComponent } from '@/components/Scoreboard';
import { actions } from '@/modules/GameWithRedux';

export const Scoreboard: FC = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const dispatch = useDispatch();

  useEffect(() => {
    const urlLevelParam = (searchParams.get('level') ||
      undefined) as LevelNames;
    if (urlLevelParam) {
      dispatch(actions.changeLevel(urlLevelParam as LevelNames));
    }
  }, []);

  const { level, time, bombs, flagCounter } = useSelector(
    ({ game: { level, time, bombs, flagCounter } }: RootState) => ({
      level,
      time,
      bombs,
      flagCounter,
    })
  );

  const onChangeLevel = useCallback(
    ({ target: { value: level } }: React.ChangeEvent<HTMLSelectElement>) => {
      setSearchParams({ level });
      dispatch(actions.changeLevel(level as LevelNames));
    },
    // Stryker disable next-line ArrayDeclaration
    []
  );

  const onReset = useCallback(
    () => dispatch(actions.reset()),
    // Stryker disable next-line ArrayDeclaration
    []
  );

  return (
    <ScoreboardComponent
      time={String(time)}
      bombs={String(bombs - flagCounter)}
      levels={GameLevels as unknown as string[]}
      defaultLevel={level}
      onChangeLevel={onChangeLevel}
      onReset={onReset}
    />
  );
};
