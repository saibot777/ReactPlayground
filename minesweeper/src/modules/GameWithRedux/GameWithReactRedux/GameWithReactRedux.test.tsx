import React from 'react';
import { Provider } from 'react-redux';
import { useSearchParams } from 'react-router-dom';

import { render } from '@testing-library/react';

import { store } from '@/store';

import { GameWithReactRedux } from './GameWithReactRedux';

jest.mock('react-router-dom', () => ({
  useSearchParams: jest.fn(),
}));

(useSearchParams as jest.Mock).mockReturnValue([
  { get: () => null },
  jest.fn(),
]);

it('GameWithReactRedux renders correctly', () => {
  const { asFragment } = render(
    <Provider store={store}>
      <GameWithReactRedux />
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});
