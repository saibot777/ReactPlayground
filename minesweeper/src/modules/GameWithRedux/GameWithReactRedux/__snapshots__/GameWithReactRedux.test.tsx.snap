// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameWithReactRedux renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.emotion-1 {
  display: inline-block;
  padding: 0 0.3vw;
  color: #ec433c;
  border: 0.15vw inset;
  line-height: 2vw;
  letter-spacing: 0.08em;
  background: #333;
  text-shadow: 0 0 0.1vw #ec433c;
}

.emotion-2 {
  margin: 0;
  height: 100%;
  border-radius: 0;
  border: 0.15vw solid;
  border-color: white #9e9e9e #9e9e9e white;
  background-color: #d1d1d1;
}

.emotion-3 {
  font-weight: normal;
  display: block;
  white-space: nowrap;
  min-height: 1.2em;
  padding: 0 0.2vw 0.2vw;
}

.emotion-6 {
  font-size: 1.1vw;
  height: 100%;
  cursor: pointer;
  font-weight: 700;
  border-width: 0.15vw;
  border-style: solid;
  background-color: #d1d1d1;
  border-color: white #9e9e9e #9e9e9e white;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      0
    </div>
    <div>
      <select
        class="emotion-2"
      >
        <option
          class="emotion-3"
          value="beginner"
        >
          beginner
        </option>
        <option
          class="emotion-3"
          value="intermediate"
        >
          intermediate
        </option>
        <option
          class="emotion-3"
          value="expert"
        >
          expert
        </option>
      </select>
      <button
        class="emotion-6"
      >
        ðŸ™‚
      </button>
    </div>
    <div
      class="emotion-1"
    >
      10
    </div>
  </div>
  .emotion-0 {
  display: grid;
  grid-template-columns: repeat(9, auto);
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  padding: 1vw;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  width: 1.8vw;
  height: 1.8vw;
  color: transparent;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;
}

.emotion-1:hover {
  -webkit-filter: brightness(1.1);
  filter: brightness(1.1);
}

<div
    class="emotion-0"
    role="grid"
    size="9"
  >
    <div
      class="emotion-1"
      data-testid="0,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="0,1"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="0,2"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="0,3"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="0,4"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="0,5"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="0,6"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="0,7"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="0,8"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,1"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,2"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,3"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,4"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,5"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,6"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,7"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="1,8"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="2,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="2,1"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="2,2"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="2,3"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="2,4"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="2,5"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="2,6"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="2,7"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="2,8"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="3,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="3,1"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="3,2"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="3,3"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="3,4"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="3,5"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="3,6"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="3,7"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="3,8"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="4,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="4,1"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="4,2"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="4,3"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="4,4"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="4,5"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="4,6"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="4,7"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="4,8"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="5,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="5,1"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="5,2"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="5,3"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="5,4"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="5,5"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="5,6"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="5,7"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="5,8"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="6,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="6,1"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="6,2"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="6,3"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="6,4"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="6,5"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="6,6"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="6,7"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="6,8"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="7,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="7,1"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="7,2"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="7,3"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="7,4"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="7,5"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="7,6"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="7,7"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="7,8"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="8,0"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="8,1"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="8,2"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="8,3"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="8,4"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="8,5"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="8,6"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="8,7"
      role="cell"
    >
      10
    </div>
    <div
      class="emotion-1"
      data-testid="8,8"
      role="cell"
    >
      10
    </div>
  </div>
</DocumentFragment>
`;
